src/neptune/common/backends/utils.py

```diff
@@ -95,14 +95,22 @@ def wrapper(*args, **kwargs):
                 HTTPServiceUnavailable,
                 HTTPGatewayTimeout,
                 HTTPBadGateway,
-                HTTPTooManyRequests,
                 HTTPInternalServerError,
                 NewConnectionError,
                 ChunkedEncodingError,
             ) as e:
                 time.sleep(min(2 ** min(10, retry), MAX_RETRY_TIME))
                 last_exception = e
                 continue
+            except HTTPTooManyRequests as e:
+                response_headers = e.response.json()["headers"]
+                if "retry-after" in response_headers:
+                    retry_after = int(response_headers["retry-after"][0])
+                    time.sleep(min(retry_after, MAX_RETRY_TIME))
+                else:
+                    time.sleep(min(2 ** min(10, retry), MAX_RETRY_TIME))
```

Comment:
Not sure if it works for you but this could be probably simplified to something like:
```suggestion
                wait_time = int(headers["retry-after"][0]) if "retry-after" in headers else 2 ** min(10, retry)
                time.sleep(min(wait_time, MAX_RETRY_TIME))
```
And we should probably introduce some try-catching here to make the standard retry will be default even is case that we've introduced some breaking changes in the API.

Reply:
This is pretty much how much it works, I changed it cause your solution looks much better!
Also, added try-catches, good catch 😄

Reply:
This is pretty much how much it works, I changed it cause your solution looks much better!
Also, added try-catches, good catch 😄